roachtest: add slow disk perturbation test

andrewbaptist opened the pull request:
This change adds a new set of perturbation tests perturbation/*/slowDisk which tests slow disks. We have see support cases where slow disks can cause cluster level availability outages.

Epic: none

Release note: None

Commit:
roachtest: add Slow to DiskStaller to allow slowing

Previously the disk staller could only stall or unstall disks. In some
instances having a slow disk is worse than a stalled disk. This change
adds the ability to test a slow disk. This also adds a NoopDiskStaller
which does not do anything.

Epic: none

Release note: None

Commit:
roachtest: add slow disk perturbation test

This change adds a new set of perturbation tests perturbation/*/slowDisk
which tests slow disks. We have see support cases where slow disks can
cause cluster level availability outages.

Epic: none

Release note: None

Commit:
roachtest: allow pointers for perturbation test names

Previously the perturbation tests required the perturbation to be a
struct and not a pointer to a struct. This change allows the name to be
a pointer to a struct as well. This does not change the behavior of any
of the pertubation/* tests.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132129)
<!-- Reviewable:end -->

Commit:
roachprod: add WAL failover to perturbation tests

Update the perturbation/*/slowDisk tests to add WAL failover. Make this
the default for the full test.

Epic: none

Release note: None

Commit:
roachtest: add metamorphic parameters to perturbation tests

Previously all the metamorphic parameters were defined at the test
level. For a number of perturbations there are test specific parameters.
This commit allows specifying per-test perturbations and adds reasonable
defaults for all the full and dev tests.

Epic: none

Release note: None

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-:-O8mHyuG3gXmmVDePZgO:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @herkolategan, and @renatolabs)

___
*[`pkg/cmd/roachtest/roachtestutil/disk_stall.go` line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mH3_G7oA3YIw_-xny:-O8mH3_G7oA3YIw_-xnz:bryo5ks) ([raw file](https://github.com/cockroachdb/cockroach/blob/40bc44fbf646df786f08e11e55949379053ceb58/pkg/cmd/roachtest/roachtestutil/disk_stall.go#L84)):*
> ```Go
> 
> func (s *cgroupDiskStaller) Stall(ctx context.Context, nodes option.NodeListOption) {
> 	s.Slow(ctx, nodes, 4)
> ```

Do you know why we set this to 4 bytes previously, obviously not directly related to this PR but curious if there's a reason for it.
___
*[`pkg/cmd/roachtest/tests/admission_control_latency.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mHQr3AO6Fga6Q7lqq:-O8mHQr3AO6Fga6Q7lqr:b3tupu) ([raw file](https://github.com/cockroachdb/cockroach/blob/829b496b2426a4952113114a389fa5cf237bc4c2/pkg/cmd/roachtest/tests/admission_control_latency.go#L208)):*
> ```Go
> 	addMetamorphic(r, decommission{}, 3.0)
> 	addMetamorphic(r, backfill{}, 40.0)
> 	addMetamorphic(r, &slowDisk{}, math.Inf(1))
> ```

IIUC -- this is disabling the pass/fail criteria for the tests? SGTM if so, we still get some coverage by throttling the disk to check for crash bugs etc.
___
*[`pkg/cmd/roachtest/tests/admission_control_latency.go` line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mHaHTDitg_CoJmRFD:-O8mHaHTDitg_CoJmRFE:b-j3s5vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/829b496b2426a4952113114a389fa5cf237bc4c2/pkg/cmd/roachtest/tests/admission_control_latency.go#L379)):*
> ```Go
> 		s.staller = roachtestutil.NoopDiskStaller{}
> 	} else {
> 		s.staller = roachtestutil.MakeCgroupDiskStaller(t, v, false, false)
> ```

nit: add the field names e.g. `false /* ... */`


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-:-O8qVbW7CHXLdUXTDgVU:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @kvoli, and @renatolabs)

___
*[`pkg/cmd/roachtest/roachtestutil/disk_stall.go` line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mH3_G7oA3YIw_-xny:-O8qTnbX643vkRYNxwnE:b8g0zjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/40bc44fbf646df786f08e11e55949379053ceb58/pkg/cmd/roachtest/roachtestutil/disk_stall.go#L84)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Do you know why we set this to 4 bytes previously, obviously not directly related to this PR but curious if there's a reason for it.
</blockquote></details>

This was an attempt to fully stall the disk, however a 0 value was interpreted as unset, so 4 was the "lowest allowed" value. There was a comment below that I moved up here.
___
*[`pkg/cmd/roachtest/tests/admission_control_latency.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mHQr3AO6Fga6Q7lqq:-O8qU2LKDs4_ykZExUQu:b-n6yah1) ([raw file](https://github.com/cockroachdb/cockroach/blob/829b496b2426a4952113114a389fa5cf237bc4c2/pkg/cmd/roachtest/tests/admission_control_latency.go#L208)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

IIUC -- this is disabling the pass/fail criteria for the tests? SGTM if so, we still get some coverage by throttling the disk to check for crash bugs etc.
</blockquote></details>

Correct - we will still get the coverage and the runs happening with the ability to see how much we improve.
___
*[`pkg/cmd/roachtest/tests/admission_control_latency.go` line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132129#-O8mHaHTDitg_CoJmRFD:-O8qVPIBEYcS9s3iAVZw:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/829b496b2426a4952113114a389fa5cf237bc4c2/pkg/cmd/roachtest/tests/admission_control_latency.go#L379)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: add the field names e.g. `false /* ... */`
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
TFTR!

bors r=kvoli
